### ISSUE TITLE
OPENSTREAM on XNS File Server (Dodo) no longer works (as of Nov 20, 2023)

### ISSUE THREAD

**Issue by fghalasz** [Interlisp/medley#1730]
@lmm and @rmkaplan:  Looks like PRs from last year and from July 2022 have come back to bite us in the NSFILING code.

In current master (and any release after Nov 20, 2023), OPENSTREAM on an XNS file server no longer works.

`(OPENSTREAM '{DUNLIN%:}<medley>test23 'OUTPUT 'NEW)` breaks with "ARG NOT FDEV".    Error occurs in \NSRANDOM.CREATE.STREAM, specifically in the call `(create STREAM ...)`.   

More specifically, the DATATYPE STREAM definition has a CREATE clause that attempts to set the EXTERNALFORMAT field with data based on the DEVICE field.  Problem arises because in this call to (create STREAM ...), DEVICE is NIL.

Most immediately, this bug was activated in PR#1417 (Recompile files that seemed to have \IS.NO.RANDACCESSP in compiled code) when NSFILING was recompiled with no source code changes.  This happened on Nov 20, 2023.

But the root of the bug was introduced in July 2022 in PR#814 (Rmk51 end game of external format integration) when the aforementioned CREATE clause was added to the DATATYPE STREAM definition.

I can "fix" the immediate XNS OPENSTREAM problem by bypassing the call to \EXTERNALFORMAT (in the DATATYPE STREAM CREATE clause) if the DEVICE field is NIL.   But since I don't understand the whole \EXTERNALFORMAT scheme, I am not sure that this is a true fix or would cause problems elsewhere in the NSFILING code.

**Comment by rmkaplan** [comment]
What kind of stream is it, that doesn’t have a device?  What can it do?

If this is a reasonable situation, then a reasonable fix is to make the change you propose, to make the CREATE clause in the stream declaration test for the device and fall through to :DEFAULT if it doesn’t have one.   (:DEFAULT currently maps to :XCCS).  But if it doesn’t have a device, I’m not sure that it matters whether or not it has a format.

**Comment by fghalasz** [comment]
@rmkaplan  This happens during OPENSTREAM on an XNS File Server file.  The NSFILING code first creates a skeleton stream (without specifying a DEVICE field) and then later in fills in the DEVICE field (along with other fields).  The break happens when it attempts to create the skeleton stream.

But I fixed it by reworking the NSFILING code a bit - a DEVICE was available when the skeleton stream is being created - it just wasn't being passed down to the function that was doing the skeleton creation.  Added a DEVICE argument to that function and added the DEVICE initialization to the create STREAM.

Fix is in PR#1733.
